<a href="https://colab.research.google.com/github/JosenildoJunior/StatPyDataScience/blob/main/Python_b%C3%A1sico.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Estatística com Python: Um Guia para Estudos e Solução de Problemas


## Neste módulo, vamos aprender sobre os conceitos básicos de Python, incluindo:

* **Variáveis:** Como declarar e usar variáveis para armazenar dados.
* **Tipos de dados:** Os diferentes tipos de dados que podem ser armazenados em variáveis.
* **Estruturas:** São objetos que armazenam e organizam dados.
* **Operadores e operações matemáticas:** Como realizar operações matemáticas e lógicas.
* **Estruturas de controle:** Como controlar o fluxo de execução de seu código.
* **Pandas:** Biblioteca para manipulação e análise de dados.

## **Variáveis**

Uma variável é um local na memória do computador destinado ao armazenamento de valores. Para ilustrar esse conceito, considere o exemplo de armazenar a idade de uma pessoa, como 27 anos, em uma variável. No contexto do Python, como poderíamos realizar essa operação?

In [ ]:
# Criando a variável
idade = 27

Podemos observar que para criar uma variável, basta escolher um nome e, em seguida, utilizar o símbolo de igual (=), que funciona como uma atribuição. Portanto, ao usar essa abordagem, criamos uma variável chamada 'idade' e a associamos ao valor 27.

Agora que a variável foi declarada, podemos acessar seu valor a qualquer momento necessário. Utilizaremos a função 'print' para exibir o valor contido na variável.

In [ ]:
# Exibindo o valor
print(idade)

27



Dessa forma, podemos imprimir o valor da variável sempre que necessário. É importante lembrar que após ser declarada, uma variável pode ser acessada e modificada usando seu nome. Vamos alterar o valor da idade para 20.

In [ ]:
# Alterando o valor da variável
idade = 20

Para realizar essa alteração, bastou utilizarmos a mesma variável, porém, ao atribuirmos um valor diferente, ela automaticamente sobrescreveu o antigo valor que continha. Dessa forma, a variável 'idade' passou a ter o valor de 20, diferentemente da primeira vez em que a criamos, quando seu valor era 27.

In [ ]:
# Exibindo o valor
print(idade)

20


Podemos confirmar, então, que o valor da variável foi alterado. As variáveis são uma ferramenta fundamental na programação em Python, permitindo-nos armazenar dados e manipulá-los de forma eficiente.

## **Tipos de dados**

Os tipos de dados determinam o tipo de valor que pode ser armazenado em uma variável. Os tipos de dados mais comuns são:

- Inteiro: armazena números inteiros, como 1, 2, 3, etc.
- Ponto flutuante: armazena números de ponto flutuante, como 1.0, 2.5, 3.14159, etc.
- Booleano: armazena valores lógicos, como True (verdadeiro) ou False (falso).
- String: armazena texto, como "Olá, mundo!".
- Date: armazenar informações de data, como (2024, 1, 10)



### **Int (Inteiros)**

Como já vimos anteriormente, variáveis do tipo int representam números inteiros, como a idade de alguém, por exemplo. Ao declararmos uma variável e atribuirmos a ela um valor numérico inteiro, ela será automaticamente do tipo int. Vamos observar com um exemplo:

In [ ]:
# Criando uma variável do tipo int
gols = 2

Criamos uma variável chamada 'gols' e atribuímos a ela um valor inteiro. Para garantir que essa variável seja do tipo int, podemos utilizar a função chamada 'type'. Esta função retornará o tipo da variável, seja qual for.

In [ ]:
# Observando o tipo da variável
type(gols)

int

Ao observar o resultado da função 'type' e encontrar 'int', confirmamos que a variável é do tipo inteiro. Isso é uma maneira útil de verificar o tipo de variável que estamos manipulando em Python.

### **Float (Ponto flutuante)**

No tipo float, os dados armazenados representam números reais com casas decimais, como 1.0, 2.5 e 3.14159. Vamos explorar mais a fundo para uma compreensão mais ampla.

In [ ]:
# Criando a variável do tipo float
altura = 1.74

Para exemplificar, criamos uma variável que recebe a altura de um indivíduo em metros. Por se tratar de um valor real, podemos concluir que é do tipo float, mas vamos confirmar.



In [ ]:
# Observando o tipo da variável
type(altura)

float

E é exatamente isso que confirmamos ao executarmos a função `type`. Percebemos que a saída foi float, confirmando que esse dado se trata de um valor do tipo float.

### **Bool (Booleano)**


Os dados booleanos, ou bool, são dados que só podem receber dois valores: verdadeiro (True) ou falso (False). Eles são usados para representar valores lógicos, como a verdade ou falsidade de uma afirmação.

In [ ]:
# Criando a variável do tipo bool
falso = False

# Criando a variável do tipo bool
verdadeiro = True

# Criando a variável do tipo bool
casa = False

É importante frisar que o tipo da variável é independente do nome. O que realmente importa para determinar o tipo da variável é o conteúdo. Neste exemplo, foram criadas três variáveis, todas recebendo valores booleanos, que podem ser verdadeiro ou falso. Apesar de terem nomes diferentes, os conteúdos são do mesmo tipo. Portanto, ao utilizarmos a função 'type', o resultado será o mesmo. Vamos conferir.

In [ ]:
# Observando o tipo da variável
type(falso), type(verdadeiro), type(casa)

(bool, bool, bool)

Podemos observar, então, que em todos os casos o tipo de dado é o mesmo; ambas são do tipo booleano.

### **String (Texto)**

Strings são dados do tipo texto; normalmente, esse tipo de dado está entre aspas. Logo, se declararmos uma variável com um nome qualquer e seu conteúdo estiver entre aspas (sejam elas simples ou duplas), isso implica que o conteúdo daquela variável será do tipo string. Vamos ver no exemplo a seguir.

In [ ]:
# Criando a variável do tipo string
nome = 'Pepper'

Então, a nossa variável está criada, mas como posso ter certeza de que essa variável se trata realmente de uma variável do tipo texto? Para sanarmos essa dúvida, vamos utilizar novamente a função 'type'.

In [ ]:
# Observando o tipo da variável
type(nome)

str

Dessa forma, podemos ter certeza do tipo de dados armazenado em qualquer variável que utilizemos essa função. Neste caso, o resultado foi 'str' ou string, logo podemos concluir que o tipo da variável é do tipo texto.

### **Datetime (Data)**

No Python, a variável do tipo date é usada para representar datas. Esta variável faz parte do módulo datetime na biblioteca padrão do Python. Ela é usada para armazenar informações de data sem incluir informações sobre o horário específico.

Para exemplificar esse tipo de dados, será necessário a importação de uma biblioteca. Vamos observar essa parte.

In [ ]:
# Realizando a importação
from datetime import date

Agora que realizamos a importação do pacote, podemos seguir para a exemplificação.

In [ ]:
# Obter o dia de hoje
data_atual = date.today()

Atribuímos à variável data_atual o valor da função 'today()', que retorna a data atual. Vamos observar o dado armazenado e, em seguida, o tipo do dado.

In [ ]:
# Exibindo a data de hoje
print(data_atual)

2024-01-15


Dessa forma, podemos observar o valor armazenado, que é o exato dia em que estou executando este código. Agora, vamos confirmar qual o tipo de dado presente nessa variável.

In [ ]:
# Observando o tipo do dado
type(data_atual)

datetime.date

O resultado foi datetime.date, logo podemos concluir que esse dado se trata realmente de um dado do tipo data.

## **Estruturas**

As estruturas de dados são objetos que armazenam e organizam dados. Elas são usadas para simplificar o código e tornar mais eficiente o processamento de dados.

As principais estruturas de dados em Python são:

- Listas: armazenam uma coleção de valores de qualquer tipo.
- Tuplas: armazenam uma coleção de valores de qualquer tipo, mas são imutáveis.
- Dicionários: armazenam uma coleção de pares chave-valor.
- Conjuntos: armazenam uma coleção de valores únicos.

### **List (Listas)**

Listas são estruturas de dados ordenadas que permitem armazenar uma coleção de valores de qualquer tipo. Os valores de uma lista são chamados de elementos ou itens.

Para declararmos uma lista em Python, seguimos o seguinte padrão: nome_da_variável, seguido por um sinal de igual (=) e os valores entre colchetes ([ ]), como por exemplo:

In [ ]:
# Criando uma lista
frutas = ['Abacate', 'Abacate', 'Uva', 12, 3.14, 12]

Podemos notar uma coisa interessante: uma lista pode ser criada e utilizada para armazenar dados de diferentes tipos. Podemos observar que nessa mesma lista chamada frutas, temos variáveis dos tipos string, int e float. Vamos verificar se essa variável realmente se trata de uma lista.

Observando a lista criada.

In [ ]:
# Exibindo os valores
print(frutas)

['Abacate', 'Abacate', 'Uva', 12, 3.14, 12]


Agora vamos observar o tipo do dado.

In [ ]:
# Exibindo o tipo do dado
type(frutas)

list

Podemos observar então que essa variável realmente se trata de uma lista e que pode armazenar diversos tipos de dados. Para demonstrar o quão poderosa é essa estrutura de dados, vamos observar como realizar funções como, por exemplo:

- Acessar elementos
- Adicionar elementos
- Alterar elementos
- Remover elementos
- Contar elementos
- Verificando a existência de elementos
- Buscar elementos
- Ordenar a lista
- Invertendo a lista



#### *Acessando um elemento*

Para acessar um elemento em uma lista, basta passar a posição que o elemento está na lista entre colchetes. Lembrando que a contagem desses números começa em 0 no Python. Então, digamos que existe a seguinte lista: notas = [6.0, 10, 9.7] Podemos notar que temos 3 elementos. Mas, se quisermos acessar o primeiro, utilizaremos o código da seguinte forma: notas[0], Pois, como a contagem em Python começa de 0, o primeiro elemento da lista está na posição 0, o segundo na posição 1, o terceiro na posição 2 e assim por diante.

Primeiro, vamos exibir a lista completa.

In [ ]:
# Exibindo os valores
print(frutas)

['Abacate', 'Abacate', 'Uva', 12, 3.14, 12]


Agora, vamos acessar o terceiro elemento da lista.

In [ ]:
# Acessando um elemento na lista
frutas[2]

'Uva'

Ao observar o resultado, podemos ver que realmente exibimos o terceiro elemento na lista. Isso ocorre porque, em Python, o primeiro elemento de uma lista tem índice 0, o segundo elemento tem índice 1 e assim por diante. Portanto, para exibir o terceiro elemento, passamos como parâmetro o índice 2.

#### *Adicionando um elemento*

Para adicionar um elemento a uma lista, basta utilizarmos a função append(). Essa função recebe como parâmetro o item que queremos adicionar à lista. Vamos observar um exemplo:

In [ ]:
# Adicionando um elemento a lista
frutas.append("Laranja")

Agora, com o item adicionado, vamos observar a lista completa.

In [ ]:
# Exibindo os valores
print(frutas)

['Abacate', 'Abacate', 'Uva', 12, 3.14, 12, 'Laranja']


Podemos então notar que o elemento foi adicionado com sucesso à lista.

#### *Alterando elementos*

Para alterar um elemento de uma lista, basta indicarmos o índice do elemento que queremos alterar e, em seguida, passar o novo valor para esse índice. Vamos observar um exemplo:

In [ ]:
# Alterando um elemento
frutas[4] = 'Maracúja'

Assim, realizamos uma alteração de um elemento na lista. Vamos observar como ela está agora.

In [ ]:
# Exibindo os valores
print(frutas)

['Abacate', 'Abacate', 'Uva', 12, 'Maracúja', 12, 'Laranja']


Podemos confirmar que o elemento que desejamos foi alterado no índice que queríamos.

#### *Removendo elementos*

Existem diferentes métodos para remover um elemento de uma lista. Dois dos mais comuns são:

- O método remove( ) remove o primeiro elemento da lista que corresponder ao valor especificado.
- O método pop( ) remove o elemento da lista na posição especificada.

Vamos observar primeiro o método remove.

In [ ]:
# Removendo um elemento
frutas.remove(12)

Agora que removemos um elemento, vamos observar como está a lista atual.

In [ ]:
# Exibindo o resultado
print(frutas)

['Abacate', 'Abacate', 'Uva', 'Maracúja', 12, 'Laranja']


Dessa forma, podemos observar que realmente conseguimos remover o elemento 12 da lista. No entanto, essa função removeu apenas o primeiro valor 12 que encontrou. Dessa forma, o segundo elemento 12 continua na lista. Vamos utilizar agora o método pop( ) para removermos um elemento com base em sua posição na lista.

In [ ]:
# Removendo um elemento
frutas.pop(4)

12

Agora que removemos todos os elementos que desejamos, vamos observar como ficou a nossa lista.

In [ ]:
# Exibindo os valores
print(frutas)

['Abacate', 'Abacate', 'Uva', 'Maracúja', 'Laranja']


Agora sim, removemos e alteramos todos os valores que desejávamos na nossa lista, ficando apenas com 5 elementos.

#### *Contando os elementos*

Vamos utilizar o método len() para contarmos quantos elementos temos na nossa lista. Utilizaremos esse método da seguinte forma:

In [ ]:
# Exibindo o número de elementos
len(frutas)

5

Como vimos anteriormente, nossa lista realmente possui apenas 5 elementos.

Também é possível contarmos quantas vezes um mesmo elemento aparece na lista. Para isso, utilizaremos o método count().

In [ ]:
# Contando quantas vezes o mesmo elemento aparece
frutas.count('Abacate')

2

Quando pesquisamos pelo elemento 'Abacate', é possível notar que ele retorna o número 2, que é exatamente o número de elementos chamados 'Abacate' que temos na lista. Vamos pesquisar agora o elemento 'Uva' para ver o que encontramos.

In [ ]:
# Contando quantas vezes o mesmo elemento aparece
frutas.count('Uva')

1

Exatamente isso que acontece. Na nossa lista, existem 2 abacates e apenas uma uva.

#### *Verificando se existe um determinado elemento*

Para verificar se um elemento existe em uma lista, podemos utilizar o operador 'in'. Ele retorna True se o elemento estiver presente na lista e False caso contrário.

Primeiro, vamos exibir a lista para observarmos os elementos existentes nela.

In [ ]:
# Exibindo os valores
print(frutas)

['Abacate', 'Abacate', 'Uva', 'Maracúja', 'Laranja']


Agora, vamos verificar se o elemento 'Uva' está na lista.

In [ ]:
# Buscando um elemento na lista
"Uva" in frutas

True

O resultado foi True, logo podemos afirmar que existe um elemento chamado Uva na nossa lista. Vamos tentar pesquisar algum elemento que não existe.

In [ ]:
# Buscando um elemento na lista
"Manga" in frutas

False

Ao verificarmos o elemento chamado Manga, recebemos um resultado False. Logo, podemos afirmar que esse elemento não está contido na lista.

#### *Buscando elementos*

Se quisessemos descobrir a posição que determinado elemento ocupa na lista iremos utilizar o método 'index()'.

Vamos exibir a lista completa.

In [ ]:
# Exibindo os valores
print(frutas)

['Abacate', 'Abacate', 'Uva', 'Maracúja', 'Laranja']


Agora, vamos aplicar o método para buscar um elemento na lista.

In [ ]:
# Buscando um elemento
frutas.index("Abacate")

0

Como a lista contém dois abacates, a função retornou a posição do primeiro. Vamos observar outro exemplo.

In [ ]:
# Buscando um elemento
frutas.index("Maracúja")

3

Como podemos observar, o elemento Maracujá ocupa o segundo lugar na nossa lista. Logo, o seu índice é o número 1. Se buscássemos um elemento que não existe na lista, o que aconteceria?

In [ ]:
# Buscando um elemento
frutas.index("Manga")

ValueError: 'Manga' is not in list

Podemos observar que, ao tentarmos buscar um elemento que não existe na lista com esse método, ocorrerá um erro, informando que o elemento que estamos procurando não existe na lista.

#### *Ordenando a lista*

Para ordenar a lista, utilizaremos o método sort(). Com esse método, é possível ordenar a lista tanto em ordem crescente quanto em ordem decrescente. Primeiramente, vamos observar como ordenar em ordem crescente.

Primeiro, vamos observar a atual ordenação da nossa lista.

In [ ]:
# Exibe os valores
print(frutas)

['Abacate', 'Abacate', 'Uva', 'Maracúja', 'Laranja']


Ordenando a lista em ordem crescente.

In [ ]:
# Ordena a lista em ordem crescente
frutas.sort()

Agora que ordenamos a lista, vamos observar como ela está.

In [ ]:
# Exibe os valores
print(frutas)

['Abacate', 'Abacate', 'Laranja', 'Maracúja', 'Uva']


Podemos confirmar que agora nossa lista está em ordem crescente. Vamos fazer ao contrário e deixar a lista em ordem decrescente.

In [ ]:
# Ordena a lista em ordem decrescente
frutas.sort(reverse=True)

Vamos observar como está nossa lista agora.

In [ ]:
# Exibe os valores
print(frutas)

['Uva', 'Maracúja', 'Laranja', 'Abacate', 'Abacate']


Podemos, então, observar como podemos alterar a ordem da nossa lista.

#### *Invertendo a ordem dos elementos*

Vamos observar como podemos inverter a ordem dos elementos presentes na lista. Para isso, utilizaremos o método reverse( ).

Novamente, vamos observar como está a lista atualmente.

In [ ]:
# Exibindo os valores
print(frutas)

['Uva', 'Maracúja', 'Laranja', 'Abacate', 'Abacate']


Agora, vamos verificar como podemos inverter a lista na prática.

In [ ]:
# Inverte a ordem dos elementos na lista
frutas.reverse()

Agora que já aplicamos o método de inversão, vamos observar como está a nossa lista.

In [ ]:
# Exibindo os valores
print(frutas)

['Abacate', 'Abacate', 'Laranja', 'Maracúja', 'Uva']


Como podemos observar, nossa lista voltou para a ordem inicial.

### **Tuple (tuplas)**

Tuplas são estruturas de dados ordenadas que permitem armazenar uma coleção de valores de qualquer tipo. No entanto, tuplas são imutáveis, o que significa que não é possível alterar seus elementos após a criação.

As tuplas possuem certas semelhanças com as listas, porém é importante frisar que as listas são mutáveis enquanto as tuplas são imutáveis. Vamos observar como declarar uma tupla.

In [ ]:
# Criando uma tupla
vegetais = ("abóbora", "pimentão", "chuchu", "pimentão")

Aqui podemos observar que a criação de uma tupla é bem semelhante à criação de uma lista, mas levando em consideração que os elementos da tupla não podem ser alterados, normalmente os elementos que são guardados nelas são exatamente elementos que não precisam ser alterados, como dados históricos por exemplo.




Vamos observar algumas funções que podemos utilizar nas tuplas:

- len(): retorna o comprimento da tupla.
- in: verifica se um elemento está presente na tupla.
- index(): retorna o índice de um elemento na tupla.
- count(): retorna o número de ocorrências de um elemento na tupla.
- sorted(): retorna uma nova tupla ordenada.

Essas funções são bem semelhantes com as que utilizamos na lista. Vamos observar algumas dessas na prática.

#### *Acessando um elemento*

Antes de acessarmos um elemento, vamos observar a tupla completa.

In [ ]:
# Exibindo os valores
print(vegetais)

('abóbora', 'pimentão ', 'chuchu', 'pimentão')



Agora vamos observar como acessar um elemento de acordo com o seu índice.

In [ ]:
# Acessando um elemento na tupla
vegetais[1]

'pimentão '


Podemos observar que a maneira de acessar um elemento tanto na lista quanto na tupla são bem semelhantes. Utilizamos praticamente a mesma lógica: passamos o nome da tupla e, em seguida, entre colchetes, o índice do elemento que queremos observar.

#### *Contando os elementos*

Vamos utilizar o método len( ) para contarmos quantos elementos temos na nossa tupla.

In [ ]:
# Exibindo o número de elementos
len(vegetais)

4

Dessa forma, podemos confirmar que a nossa tupla possui pelo menos 4 elementos.

Agora, vamos utilizar o método count( ) para sabermos quantas vezes existe um determinado elemento na nossa tupla.

In [ ]:
# Contando quantas vezes o mesmo elemento aparece
vegetais.count("pimentão")

2

Ao conferirmos quantos elementos "pimentão" temos na tupla, o valor retornado foi o número 2. Isso indica que o elemento "pimentão" aparece 2 vezes na tupla.


Vamos tentar contar um elemento que não existe na tupla.

In [ ]:
# Contando quantas vezes o mesmo elemento aparece
vegetais.count(3)

0

Como pesquisamos um valor que não existe na tupla, recebemos um valor igual a 0, que está condizente com a realidade da nossa tupla.

#### *Verificando se existe um determinado elemento*

Para isso, utilizaremos o operador "in" da seguinte forma.

In [ ]:
# Buscando um elemento na lista
'chuchu' in vegetais

True

Ao pesquisar o elemento "chuchu" na tupla, recebemos um True. Ou seja, é verdade que existe um elemento "chuchu" na tupla. Vamos verificar um exemplo onde isso não é verdade.

In [ ]:
# Buscando um elemento na lista
'morango' in vegetais

False

Como o elemento "morango" não existe na tupla, recebemos um False da nossa busca.

#### *Buscando um elemento*

Se quiséssemos descobrir a posição que determinado elemento ocupa na tupla, iremos utilizar o método 'index( )', assim como fazemos na lista.

Vamos exibir a tupla completa.

In [ ]:
# Exibindo os valores
print(vegetais)

('abóbora', 'chuchu', 2, 2, 5)


Agora, vamos aplicar o método para buscar um elemento na tupla.

In [ ]:
# Buscando um elemento
vegetais.index("abóbora")

0

Podemos observar então que o elemento "abóbora" está na posição 0 da nossa tupla.

Caso tentarmos buscar um elemento que não existe, iremos receber um erro, assim como na lista. Vamos observar um exemplo.

In [ ]:
# Buscando um elemento
vegetais.index("Tomate")

ValueError: tuple.index(x): x not in tuple

Como havíamos dito anteriormente, ao tentarmos buscar um elemento que não existe, receberemos um erro nos informando que o elemento buscado não existe na tupla.

#### *Ordenando a tupla*

Vamos utilizar o método 'sorted( )' para ordenar os elementos da tupla.

Mas primeiro, vamos observar como está ordenada a nossa tupla atualmente.

In [ ]:
# Exibindo os valores
print(vegetais)

('abóbora', 'pimentão ', 'chuchu', 'pimentão')


Agora, vamos aplicar o método 'sorted( )' para ordenar a mesma.

In [ ]:
# Ordenando a tupla
print(sorted(vegetais))

['abóbora', 'chuchu', 'pimentão', 'pimentão ']


Dessa forma, podemos notar como podemos alterar a ordem dos elementos da nossa tupla.

### **Dict (dicionário)**

Dicionários são estruturas de dados que permitem armazenar uma coleção de pares chave-valor. As chaves são usadas para acessar os valores associados a elas.

Para declararmos um dicionario no Python será feito da seguinte forma:

In [ ]:
# Criando um dic
alunos = {"Lerma": 7.0, "Ana": 9.0, "Maria": 8.0, "IA": 10.0}

Podemos observar que, na declaração, basta utilizarmos chaves. Como um dicionário funciona na estrutura chave-valor, quando preenchemos, basta declararmos os mesmos. Nesse exemplo, temos os nomes de alunos e suas respectivas notas.

Algumas das funções que podem ser utilizadas em dicionários incluem:

- len(): retorna o comprimento do dicionário.
- in: verifica se uma chave está presente no dicionário.
- get(): retorna o valor associado a uma chave, ou um valor padrão se a chave não estiver presente.
- update(): adiciona ou atualiza os valores associados a chaves existentes.
- keys(): retorna uma lista com as chaves do dicionário.
- values(): retorna uma lista com os valores do dicionário.
- items(): retorna uma lista de tuplas, cada uma contendo uma chave e seu valor associado.

Essas são algumas das funções que podemos utilizar nos dicionários. Vamos observá-las na prática.

#### *Acessando um elemento*

Para acessarmos um elemento, basta passarmos a chave do elemento que queremos acessar. Por exemplo, se quisessemos verificar a nota do aluno chamado "Lerma".

In [ ]:
# Acessando o valor com base na chave
alunos["Lerma"]

7.0

Dessa forma, sempre poderemos verificar o valor de alguma chave. Nesse caso, as nossas chaves são o nome dos alunos e os valores são suas notas.



#### *Adicionando um elemento*

Vamos utilizar o método update( ) para adicionarmos mais um elemento ao dicionário. Primeiro, vamos visualizar o dicionário completo.

In [ ]:
# Exibindo os valores
print(alunos)

{'Lerma': 7.5, 'Ana': 9.0, 'Maria': 8.0, 'IA': 10.0}


Agora, vamos adicionar mais um elemento.

In [ ]:
# Adicionando mais um elemento
alunos.update({"Ia 2.0": 9.2})

Vamos observar como ficou nosso dicionário após essa inclusão.

In [ ]:
# Exibindo os valores
print(alunos)

{'Lerma': 7.5, 'Ana': 9.0, 'Maria': 8.0, 'IA': 10.0, 'Ia 2.0': 9.2}


Podemos então observar que conseguimos adicionar um novo elemento ao dicionário.

#### *Alterando um elemento*

Nesse caso, vamos alterar a nota do aluno "Lerma". Vamos primeiro visualizar o dicionário completo.

In [ ]:
# Exibindo os valores
print(alunos)

{'Lerma': 7.0, 'Ana': 9.0, 'Maria': 8.0, 'IA': 10.0}


Agora, vamos aplicar o método update( ) para realizar uma alteração de valores no dicionário.

In [ ]:
# Atualiza o valor associado à chave "idade"
alunos.update({"Lerma": 7.5})

Vamos observar como ficou nosso dicionário agora.

In [ ]:
# Exibindo os valores
print(alunos)

{'Lerma': 7.5, 'Ana': 9.0, 'Maria': 8.0, 'IA': 10.0}


Dessa forma, podemos notar que realizamos com sucesso uma alteração na nota do aluno.

In [ ]:
type(alunos)

dict

#### *Excluindo um elemento*

Existem diversas maneiras de excluir um elemento de um dicionário. Nesse caso, vamos utilizar o método pop() para realizar a exclusão.

Vamos relembrar como está atualmente nosso dicionário.

In [ ]:
# Exibindo os valores
print(alunos)

{'Lerma': 7.5, 'Ana': 9.0, 'Maria': 8.0, 'IA': 10.0, 'Ia 2.0': 9.2}


Agora, vamos excluir o aluno "Ia 2.0".

In [ ]:
alunos.pop("Ia 2.0")

9.2

Agora que realizamos a exclusão, vamos observar o dicionário completo.

In [ ]:
# Exibindo os valores
print(alunos)

{'Lerma': 7.5, 'Ana': 9.0, 'Maria': 8.0, 'IA': 10.0}


Podemos, então, confirmar a exclusão do elemento que não queríamos.

#### *Contando os elementos*

Utilizaremos novamente o método len() para conferirmos o tamanho do nosso dicionário.

In [ ]:
# Observando o comprimento do dicionário
len(alunos)

4

É possível notar que ele conta o número de chaves presentes em nosso dicionário, sendo assim o número de elementos está correto.

#### *Verificando se existe um determinado elemento*

Para verificar a existência de algum elemento, vamos utilizar novamente o operador "in" da seguinte forma:

In [ ]:
# Buscando um elemento
"João" in alunos

False

Dessa forma, podemos conferir se existe ou não um determinado elemento no dicionário. Caso a saída seja True, indica que o elemento buscado existe no dicionário. Caso a saída seja False, indica que o elemento buscado não existe no dicionário.

#### *Observando o valor associado a uma chave*

Para retornar o valor referente a uma determinada chave que vamos buscar, utilizaremos o método get(). Vamos observar como ele funciona na prática.

Primeiro, vamos observar o dicionário completo.

In [ ]:
# Exibindo os valores
print(alunos)

{'Lerma': 7.0, 'Ana': 9.0, 'Maria': 8.0, 'IA': 10.0}


Agora, vamos retornar o valor referente à aluna "Ana".

In [ ]:
# Retorna o valor associado à chave "nome"
print(alunos.get("Ana"))

9.0


Dessa forma, podemos observar o valor referente à chave que buscamos. Mas e se buscássemos uma chave que não existe? Vamos ver o que podemos fazer a respeito disso.

In [ ]:
# Retorna o valor padrão "Não encontrado" se a chave não estiver presente
print(alunos.get("Kobe", "Aluno não cadastrado"))

Aluno não cadastrado


Ao adicionarmos mais um parâmetro ao método get(), podemos definir um valor padrão caso a busca que realizamos não seja concluída por inexistência da chave buscada. Caso buscássemos por uma chave existente, esse método retornaria o valor referente à chave. Como a chave não foi encontrada, o valor retornado foi a mensagem padrão que declaramos no mesmo método.

#### *Observando as chaves*

Utilizando o método keys(), podemos observar todas as chaves presentes no nosso dicionário. Vamos observar um exemplo.

In [ ]:
# Retornando todas as chaves
alunos.keys()

dict_keys(['Lerma', 'Ana', 'Maria', 'IA'])

Utilizando esse método, sempre será possível obter uma lista com as chaves do dicionário.

#### *Observando os valores*

Para observarmos todos os valores presentes no nosso dicionário, vamos utilizar o método values() da seguinte forma:

In [ ]:
# Retornando todos os valores
alunos.values()

dict_values([7.0, 9.0, 8.0, 10.0])

Assim, podemos sempre obter uma lista com todos os valores do dicionário.

#### *Observando tanto chave quanto o valor*

á sabemos como visualizar cada um desses separadamente. Mas, se quisemos observar tanto a chave quanto o valor de uma só vez, como poderíamos fazer? Para conseguirmos observar cada chave e seu valor associado, podemos utilizar o método items(). Vamos observar na prática.

In [ ]:
# Observando todas as chaves e seus valores
alunos.items()

dict_items([('Lerma', 7.0), ('Ana', 9.0), ('Maria', 8.0), ('IA', 10.0)])

Dessa forma, podemos observar uma lista de tuplas, onde cada tupla contém uma chave e seu valor associado.

## **Operadores**

Existem dois tipos de operadores no Python: os operadores de comparação e os operadores lógicos. Vamos observar ambos um pouco mais afundo.

### *Operadores de comparação*

Os operadores de comparação são usados para comparar dois valores e retornar um valor booleano (True ou False). Eles são usados em instruções condicionais para fazer verificações e tomar decisões.

Vamos observar quais são esses operadores de comparação:

- O operador de igualdade == é usado para verificar se dois valores são iguais.
- O operador de desigualdade != é usado para verificar se dois valores são diferentes.
- O operador de maior que > é usado para verificar se o primeiro valor é maior que o segundo valor.
- O operador de maior ou igual a >= é usado para verificar se o primeiro valor é maior ou igual ao segundo valor.
- O operador de menor que < é usado para verificar se o primeiro valor é menor que o segundo valor.
- O operador de menor ou igual a <= é usado para verificar se o primeiro valor é menor ou igual ao segundo valor.

Vamos observar um exemplo de cada um desses operadores.

**Igualdade**

In [ ]:
# Igualdade
16 == 17

False

Como os valores são diferentes, obtivemos um resultado falso.

**Desigualdade**

In [ ]:
# Desigualdade
'maçã' != 'Maçã'

True

Os valores são diferentes porque a primeira palavra começa com letra maiúscula e o Python é sensível a esse tipo de coisa. Logo, esse detalhe faz com que, aos olhos do Python, sejam elementos diferentes um do outro.

**Maior que**

In [ ]:
# Maior que
57 > 20

True

O primeiro valor é maior que o segundo, logo obtivemos um valor True.

**Maior ou igual a**

In [ ]:
# Maior ou igual
57 >= 57

True

57 é igual a 57, portanto, o resultado que obtivemos foi um True.

**Menor que**

In [ ]:
# Menor que
2289 < 2288

False

O primeiro valor é maior ou igual ao segundo, portanto, obtivemos um valor False.

**Menor ou igual a**

In [ ]:
# Menor ou igual
2289 <= 2289

True

Mesmo que o primeiro valor não seja menor que o outro, eles são iguais, por isso recebemos o valor True.

### *Operadores lógicos*

Os operadores lógicos são usados para combinar duas ou mais expressões condicionais. Eles são usados para tomar decisões baseadas em múltiplas condições.

Vamos observar quais são esses operadores lógicos:

- O operador 'and' é usado para combinar duas expressões condicionais. O resultado da expressão será True apenas se ambas as expressões forem True.

- O operador 'or' é usado para combinar duas expressões condicionais. O resultado da expressão será True se pelo menos uma das expressões for True.

- O operador 'not' é usado para inverter o valor booleano de uma expressão. Se a expressão for True, o resultado será False. Se a expressão for False, o resultado será True.


Vamos observar um exemplo de cada um desses tipos de operadores.

**And**

In [ ]:
# Aplicando o and
17 > 18 and 18 < 21

False

Como estamos utilizando o operando 'and', ambas as expressões tinham que ser verdadeiras. Como 17 é menor que 18, e na expressão estamos conferindo se, na verdade, ele é maior que 18, o resultado seria False. A segunda expressão, por outro lado, resultou em True, pois 21 é maior que 18. Logo, nossa expressão ficou da seguinte forma:

In [ ]:
# Expressão apenas com operadores
False and True

False

Como no 'and' ambas as expressões precisam ser verdadeiras, o nosso resultado foi False, pois uma das expressões teve um resultado False.

**Or**

In [ ]:
# Aplicando o or
17 > 18 or 18 < 21

True

Já no 'or', basta que um dos resultados da expressão seja verdadeiro para que o resultado seja verdadeiro. Como vimos no exemplo anterior, a expressão acima é False 'and' True, por isso nosso resultado final utilizando o operador 'or' é True.

**Or**

In [ ]:
# Aplicando o not
not(17 > 18 or 18 < 21)

False

O operador 'not' inverte o valor lógico de uma expressão. Para observarmos bem essa parte, utilizei a mesma operação que utilizamos no 'or'. Como o resultado final foi True e o operador not inverte o resultado, logo o resultado final é False, que é o contrário de True.

## **Operações matemáticas**

Os operadores matemáticos são usados para realizar operações matemáticas em Python.

**Operador de adição**

O operador de adição '+' é usado para somar dois valores, vamos observar um exemplo.

In [ ]:
# Definindo os valores das varáveis
x = 10
y = 57

# Realizando a operação
soma = x + y

# Exibindo o resultado
print(soma)

67


As operações matemáticas no Python são bastante intuitivas, então acredito que neste primeiro momento basta entendermos o seu funcionamento. Dito isso, vamos prosseguir.

**Operador de subtração**

O operador de subtração '-' é usado para subtrair dois valores.

In [ ]:
# Definindo os valores das varáveis
x = 10
y = 57

# Realizando a operação
subtracao = x - y

# Exibindo o resultado
print(subtracao)

-47


**Operador de multiplicação**

O operador de multiplicação '*' é usado para multiplicar dois valores.

In [ ]:
# Definindo os valores das varáveis
x = 10
y = 57

# Realizando a operação
multiplicacao = x * y

# Exibindo o resultado
print(multiplicacao)

570


**Operador de divisão**

O operador de divisão '/' é usado para dividir dois valores.

In [ ]:
# Definindo os valores das varáveis
x = 10
y = 57

# Realizando a operação
divisao = x / y

# Exibindo o resultado
print(divisao)

0.17543859649122806


**Operador de divisão inteira**

O operador de divisão inteira '//' é usado para realizar divisão inteira, truncando o resultado para um número inteiro.

In [ ]:
# Definindo os valores das varáveis
x = 10
y = 57

# Realizando a operação
div_inteira = x // y

# Exibindo o resultado
print(div_inteira)

0


**Operador de resto**

O operador de resto '%' é usado para calcular o resto da divisão de dois valores.

In [ ]:
# Definindo os valores das varáveis
x = 10
y = 57

# Realizando a operação
div_inteira = x % y

# Exibindo o resultado
print(div_inteira)

10


**Operador de exponenciação**

O operador de exponenciação '**' é usado para elevar um valor a um expoente.

In [ ]:
# Definindo os valores das varáveis
x = 10
y = 2

# Realizando a operação
div_inteira = x ** y

# Exibindo o resultado
print(div_inteira)

100


## **Estruturas de controle**

As estruturas de controle são instruções que controlam o fluxo de execução de um programa. Em Python, existem três tipos principais de estruturas de controle:

- Instruções condicionais: Essas instruções permitem que o fluxo de execução de um programa seja alterado com base em uma condição.
- Laços de repetição: Essas instruções permitem que um bloco de código seja executado repetidamente.

**Instruções condicionais**

As instruções condicionais são usadas para alterar o fluxo de execução de um programa com base em uma condição. Em Python, existem dois tipos de instruções condicionais:

- if-else: Esta instrução permite que o fluxo de execução de um programa seja alterado dependendo do valor de uma expressão booleana.
- elif: Esta instrução permite que o fluxo de execução de um programa seja alterado dependendo do valor de uma expressão booleana, mas apenas se a primeira expressão booleana for falsa.

Para fixarmos melhor, vamos observar um exemplo onde o if-else será utilizado. Nosso objetivo será descobrir se determinado número é par ou ímpar. Vamos observar na prática.



In [ ]:
# Definindo o número
num = 7

# Criando a nossa condição
if num % 2 == 0:
  print("Esse número é par!")
else:
  print("Esse número é impar!")

Esse número é impar!


Primeiro, declaramos qual será o número que queremos investigar. Em seguida, iniciamos a nossa condição. Para facilitar o entendimento, vamos tentar ler esse código da seguinte forma:

Se o resto da divisão de num for igual a 0, imprima a mensagem "Esse número é par!".
Senão, imprima "Esse número é ímpar!

**Laços de repetição**

Os laços de repetição são usados para executar um bloco de código repetidamente. Em Python, existem dois tipos de laços de repetição:

- for: Este laço é usado para repetir um bloco de código um número específico de vezes.
- while: Este laço é usado para repetir um bloco de código enquanto uma condição for verdadeira.

Vamos observar um exemplo do laço for, que é usado para repetir um bloco de código um número específico de vezes. A sintaxe do laço for é a seguinte:

In [ ]:
# Criando o laço
for i in range(9):
  print(i)

0
1
2
3
4
5
6
7
8


Nesse exemplo, estamos utilizando a função 'range()' para criar uma sequência de 9 números começando do 0, ou seja, basicamente vamos criar uma sequência de 0 a 9. Então, podemos ler esse código da seguinte forma:

Para todo elemento no range(0, 9), imprima cada um desses elementos.

A função 'range()' retorna uma sequência de números, começando pelo primeiro argumento e terminando antes do segundo argumento. Nesse caso, o primeiro argumento é 0 e o segundo argumento é 9. Portanto, a função 'range()' retornará uma sequência de 9 números, de 0 a 8.

O laço for irá iterar sobre essa sequência, atribuindo cada elemento à variável n. Em seguida, o bloco de código irá imprimir o valor da variável n.

Já o laço while é usado para repetir um bloco de código enquanto uma condição for verdadeira. A sintaxe do laço while é a seguinte:

In [ ]:
# Definindo o valor inicial
i = 0

# Criando o laço
while i <= 10:

  # Exibindo o valor atual
  print(i)

  # Somando mais 1 para o valor em que o looping está passando
  i += 1

  # Quando o i chegar no número 7, vamos para o looping
  if i == 8:
    break

0
1
2
3
4
5
6
7


Para entendermos melhor esse código, podemos ler ele da seguinte forma:

Primeiro, declaramos a variável inicial com o valor 0.
Em seguida, enquanto o valor de inicial for menor ou igual a 10, imprima o valor de inicial.
Logo após, somamos 1 ao valor de inicial.
Quando o valor de inicial for igual a 8, utilizamos a instrução break para pararmos a execução do programa e não gerarmos um loop infinito.

No laço while, temos duas instruções bem importantes, sendo elas:

- break: Esta instrução interrompe a execução do loop atual.
- continue: Esta instrução continua a execução do loop atual, mas pula a próxima iteração.

## **Pandas**

O pandas pode ser usado para uma variedade de tarefas de manipulação e análise de dados, incluindo:

- Leitura e escrita de dados: O pandas pode ser usado para ler e escrever dados de uma variedade de fontes, incluindo arquivos CSV, Excel, SQL e JSON.
- Manipulação de dados: O pandas pode ser usado para manipular dados, como filtrar, ordenar, agrupar e sumarizar dados.
- Análise de dados: O pandas pode ser usado para realizar análises estatísticas, como regressão, classificação e clustering.

Antes de tudo, precisamos importar a biblioteca pandas. Então, vamos fazer isso.

In [1]:
# Importando o pandas
import pandas as pd

Como, nesse caso, os meus dados estão no Drive, será necessário importar a biblioteca do Drive para que eu possa acessar o conteúdo presente no meu Drive.

In [2]:
# Acesso ao drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


Agora que já realizamos essas importações iniciais, vamos importar nossos dados para verificarmos diferentes funções do próprio pandas.

In [3]:
# Carregando os dados
df = pd.read_csv("/content/drive/MyDrive/Estatística para ciência de dados/top10s.csv", encoding="ISO-8859-1")

# Carregando outros dados
df2 = pd.read_csv("/content/drive/MyDrive/Estatística para ciência de dados/SpotifyTopSongsByCountry - May 2020.csv", encoding="ISO-8859-1")

Utilizamos a função read_csv() do pandas para realizar a leitura dos dados e armazenamos o resultado nas variáveis df e df2.

Agora, vamos entender o que cada uma das colunas presentes no conjunto de dados significa.

**Entendendo as colunas do dataset:**

* Genre - the genre of the track
* Year - the release year of the recording. Note that due to vagaries of releases, re-releases, re-issues and general madness, sometimes the release years are not what you'd expect.
* Added - the earliest date you added the track to your collection.
* Beats Per Minute (BPM) - The tempo of the song.
* Energy - The energy of a song - the higher the value, the more energtic. song
* Danceability - The higher the value, the easier it is to dance to this song.
* Loudness (dB) - The higher the value, the louder the song.
* Liveness - The higher the value, the more likely the song is a live recording.
* Valence - The higher the value, the more positive mood for the song.
* Length - The duration of the song.
* Acousticness - The higher the value the more acoustic the song is.
* Speechiness - The higher the value the more spoken word the song contains.
* Popularity - The higher the value the more popular the song is.
* Duration - The length of the song.

### **Obtendo algumas informações inicias sobre os dados**

Inicialmente, vamos observar as cinco primeiras linhas do nosso dataframe.

In [4]:
# Observando os primeiros registros do nosso df
df.head()

Unnamed: 0                 title      artist        top genre  year  bpm  \
0           1      Hey, Soul Sister       Train       neo mellow  2010   97   
1           2  Love The Way You Lie      Eminem  detroit hip hop  2010   87   
2           3               TiK ToK       Kesha        dance pop  2010  120   
3           4           Bad Romance   Lady Gaga        dance pop  2010  119   
4           5  Just the Way You Are  Bruno Mars              pop  2010  109   

   nrgy  dnce  dB  live  val  dur  acous  spch  pop  
0    89    67  -4     8   80  217     19     4   83  
1    93    75  -5    52   64  263     24    23   82  
2    84    76  -3    29   71  200     10    14   80  
3    92    70  -4     8   71  295      0     4   79  
4    84    64  -5     9   43  221      2     4   78

Agora, vamos observar as cinco últimas linhas.

In [5]:
# Observando os últimos registros
df.tail()

Unnamed: 0                                              title  \
598         599                Find U Again (feat. Camila Cabello)   
599         600      Cross Me (feat. Chance the Rapper & PnB Rock)   
600         601  No Brainer (feat. Justin Bieber, Chance the Ra...   
601         602    Nothing Breaks Like a Heart (feat. Miley Cyrus)   
602         603                                   Kills You Slowly   

               artist   top genre  year  bpm  nrgy  dnce  dB  live  val  dur  \
598       Mark Ronson   dance pop  2019  104    66    61  -7    20   16  176   
599        Ed Sheeran         pop  2019   95    79    75  -6     7   61  206   
600         DJ Khaled   dance pop  2019  136    76    53  -5     9   65  260   
601       Mark Ronson   dance pop  2019  114    79    60  -6    42   24  217   
602  The Chainsmokers  electropop  2019  150    44    70  -9    13   23  213   

     acous  spch  pop  
598      1     3   75  
599     21    12   75  
600      7    34   70  
601      1     7   69  
602      6     6   67

Agora, vamos observar os tipos das nossas variáveis.

In [6]:
# Observando os tipos das variáveis
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 603 entries, 0 to 602
Data columns (total 15 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  603 non-null    int64 
 1   title       603 non-null    object
 2   artist      603 non-null    object
 3   top genre   603 non-null    object
 4   year        603 non-null    int64 
 5   bpm         603 non-null    int64 
 6   nrgy        603 non-null    int64 
 7   dnce        603 non-null    int64 
 8   dB          603 non-null    int64 
 9   live        603 non-null    int64 
 10  val         603 non-null    int64 
 11  dur         603 non-null    int64 
 12  acous       603 non-null    int64 
 13  spch        603 non-null    int64 
 14  pop         603 non-null    int64 
dtypes: int64(12), object(3)
memory usage: 70.8+ KB


Aqui, podemos observar não só o tipo das nossas variáveis, mas também diversas outras informações sobre o nosso dataframe, como, por exemplo, o número de linhas, que é 603, o número de colunas, que é 15, e o 'Dtype', que se refere ao tipo de dado presente em cada uma das colunas do nosso dataframe.

Também conseguimos ver, em 'Non-Null Count', a quantidade de linhas não nulas presentes em cada uma das colunas do nosso dataframe.

Existe outra forma de observar exclusivamente o tipo de cada uma das colunas, que é usando 'dtypes'.

In [9]:
# Exibindo o tipo de dado presente em cada coluna
df.dtypes

Unnamed: 0     int64
title         object
artist        object
top genre     object
year           int64
bpm            int64
nrgy           int64
dnce           int64
dB             int64
live           int64
val            int64
dur            int64
acous          int64
spch           int64
pop            int64
dtype: object

Se eu quiser ver apenas colunas de um determinado tipo

In [12]:
# Observando apenas as colunas do tipo 'object'
df.dtypes[df.dtypes == 'object']

title        object
artist       object
top genre    object
dtype: object

Agora vamos observar as dimensões do nosso df

In [20]:
# Exibindo as dimensões
df.shape

(603, 15)

Aqui podemos observar o que já vimos anteriormente de que nosso df possui 603 linhas e 15 colunas

Agora vamos observar todas as colunas no nosso df

In [22]:
# Exibindo o nome das nossas colunas
df.columns

Index(['Unnamed: 0', 'title', 'artist', 'top genre', 'year', 'bpm', 'nrgy',
       'dnce', 'dB', 'live', 'val', 'dur', 'acous', 'spch', 'pop'],
      dtype='object')

Agora vamos observar algumas estatisticas sobre nosso df

In [13]:
# Resumo estátistico
df.describe()

Unnamed: 0         year         bpm        nrgy        dnce  \
count  603.000000   603.000000  603.000000  603.000000  603.000000   
mean   302.000000  2014.592040  118.545605   70.504146   64.379768   
std    174.215384     2.607057   24.795358   16.310664   13.378718   
min      1.000000  2010.000000    0.000000    0.000000    0.000000   
25%    151.500000  2013.000000  100.000000   61.000000   57.000000   
50%    302.000000  2015.000000  120.000000   74.000000   66.000000   
75%    452.500000  2017.000000  129.000000   82.000000   73.000000   
max    603.000000  2019.000000  206.000000   98.000000   97.000000   

               dB        live         val         dur       acous        spch  \
count  603.000000  603.000000  603.000000  603.000000  603.000000  603.000000   
mean    -5.578773   17.774461   52.225539  224.674959   14.326700    8.358209   
std      2.798020   13.102543   22.513020   34.130059   20.766165    7.483162   
min    -60.000000    0.000000    0.000000  134.000000    0.000000    0.000000   
25%     -6.000000    9.000000   35.000000  202.000000    2.000000    4.000000   
50%     -5.000000   12.000000   52.000000  221.000000    6.000000    5.000000   
75%     -4.000000   24.000000   69.000000  239.500000   17.000000    9.000000   
max     -2.000000   74.000000   98.000000  424.000000   99.000000   48.000000   

              pop  
count  603.000000  
mean    66.520730  
std     14.517746  
min      0.000000  
25%     60.000000  
50%     69.000000  
75%     76.000000  
max     99.000000

Aqui temos o total de valor em cada coluna, temos a média, o desvio padrão, o valor minimo, o 1, 2 e 3 quartil aleém do valor maximo

Agora vamos observar a quantidade de valores nulos

In [15]:
# Exibindo a soma de valores nulos
df.isnull().sum()

Unnamed: 0    0
title         0
artist        0
top genre     0
year          0
bpm           0
nrgy          0
dnce          0
dB            0
live          0
val           0
dur           0
acous         0
spch          0
pop           0
dtype: int64

Não temos nenhum valor nulo em nosso df

Agora vamos observar apenas os artistas unicos em nosso df

In [17]:
# Exibindo artistas únicos
df['artist'].unique()

array(['Train', 'Eminem', 'Kesha', 'Lady Gaga', 'Bruno Mars',
       'Justin Bieber', 'Taio Cruz', 'OneRepublic', 'Alicia Keys',
       'Rihanna', 'Flo Rida', 'Mike Posner', 'Far East Movement', 'Usher',
       'Sean Kingston', 'The Black Eyed Peas', 'Adam Lambert', 'Maroon 5',
       'Neon Trees', 'Selena Gomez & The Scene', 'Enrique Iglesias',
       'Katy Perry', 'Britney Spears', '3OH!3', 'David Guetta',
       'Christina Aguilera', 'Florence + The Machine', 'Shakira',
       'Tinie Tempah', 'T.I.', 'Martin Solveig', 'Christina Perri',
       'Adele', 'Pitbull', 'Beyoncé', 'Hot Chelle Rae', 'Avril Lavigne',
       'Kanye West', 'LMFAO', 'Jessie J', 'Jennifer Lopez', 'Chris Brown',
       'Sleeping At Last', 'Nicki Minaj', 'P!nk', 'Coldplay',
       'One Direction', 'Taylor Swift', 'Carly Rae Jepsen',
       'Kelly Clarkson', 'Owl City', 'The Wanted', 'fun.',
       'Ellie Goulding', 'Gym Class Heroes', 'Avicii', 'The Script',
       'Miley Cyrus', 'Swedish House Mafia', 'Daft Punk'

Dessa forma podemos observar cada um dos artistas que aparece em nosso df sem repetição

In [19]:
# Exibindo a soma artistas únicos
df['artist'].nunique()

184

Dessa forma é possivel chegarmos a conclusão de que temos 184 artistas diferentes em nosso df

Agora vamos observar os 3 menore valores da nossa coluna 'acous'

In [26]:
# Observando as 3 linhas com os valores mais baixos em 'acous'
df.nsmallest(3, 'acous')

Unnamed: 0        title     artist  top genre  year  bpm  nrgy  dnce  dB  \
3            4  Bad Romance  Lady Gaga  dance pop  2010  119    92    70  -4   
6            7     Dynamite  Taio Cruz  dance pop  2010  120    78    75  -4   
18          19    Alejandro  Lady Gaga  dance pop  2010   99    80    63  -7   

    live  val  dur  acous  spch  pop  
3      8   71  295      0     4   79  
6      4   82  203      0     9   77  
18    36   37  274      0     5   69

Agora vamos observar os 3 maiores

In [28]:
# Observando as 3 linhas com os valores mais altos em 'acous'
df.nlargest(3, 'acous')

Unnamed: 0             title       artist   top genre  year  bpm  nrgy  \
431         432             Start  John Legend  neo mellow  2016  110     4   
255         256  Not About Angels        Birdy  neo mellow  2014  116    14   
186         187             Clown  Emeli Sandé   dance pop  2013  130    23   

     dnce  dB  live  val  dur  acous  spch  pop  
431    52 -15     9   26  310     99     4   47  
255    41 -10     9   23  190     97     4   56  
186    45  -8    11   23  221     92     4   60